tests/datasets/test_streaming.py

```diff
@@ -0,0 +1,317 @@
+    }
+    return samples, decoders
+
+
+def write_synthetic_streaming_dataset(dirname: str, samples: List[Dict[str, bytes]], shard_size_limit: int) -> None:
+    first_sample_fields = list(samples[0].keys())
+    with StreamingDatasetWriter(dirname=dirname, fields=first_sample_fields,
+                                shard_size_limit=shard_size_limit) as writer:
+        writer.write_samples(samples=samples)
+
+
+@pytest.mark.parametrize("num_samples", [100, 10000])
+@pytest.mark.parametrize("shard_size_limit", [1 << 8, 1 << 16, 1 << 24])
+def test_writer(tmpdir: py.path.local, num_samples: int, shard_size_limit: int) -> None:
+    dirname = str(tmpdir)
+    samples, _ = get_fake_samples_decoders(num_samples)
+
+    first_sample_values = samples[0].values()
+    first_sample_byte_sizes = np.array([len(v) for v in first_sample_values], dtype=np.int64)
+    first_sample_bytes = len(first_sample_byte_sizes.tobytes() + b''.join(first_sample_values))
+
+    expected_samples_per_shard = shard_size_limit // first_sample_bytes
+    expected_num_shards = math.ceil(num_samples / expected_samples_per_shard)
+    expected_num_files = expected_num_shards + 1  # the index file
+
+    write_synthetic_streaming_dataset(dirname=dirname, samples=samples, shard_size_limit=shard_size_limit)
+    files = os.listdir(dirname)
+
+    assert len(files) == expected_num_files, f"Files written ({len(files)}) != expected ({expected_num_files})."
+
+
+@pytest.mark.parametrize("batch_size", [None, 1, 2])
+@pytest.mark.parametrize("share_remote_local", [False, True])
+@pytest.mark.parametrize("shuffle", [False, True])
+def test_reader(tmpdir: py.path.local, batch_size: int, share_remote_local: bool, shuffle: bool):
+    num_samples = 117
+    shard_size_limit = 1 << 8
+    samples, decoders = get_fake_samples_decoders(num_samples)
+    remote = str(tmpdir.mkdir("remote"))
+    write_synthetic_streaming_dataset(dirname=remote, samples=samples, shard_size_limit=shard_size_limit)
+    if share_remote_local:
+        local = remote
+    else:
+        local = str(tmpdir.mkdir("local"))
+
+    # Build StreamingDataset
+    dataset = StreamingDataset(remote=remote, local=local, shuffle=shuffle, decoders=decoders, batch_size=batch_size)
+
+    # Test basic sample order
+    rcvd_samples = 0
+    shuffle_matches = 0
+    for ix, sample in enumerate(dataset):
+        rcvd_samples += 1
+        uid = sample["uid"]
+        data = sample["data"]
+        expected_uid = f"{ix:06}"
+        expected_data = 3 * ix
+        if shuffle:
+            shuffle_matches += (expected_uid == uid)
+        else:
+            assert uid == expected_uid == uid, f"sample ix={ix} has uid={uid}, expected {expected_uid}"
+            assert data == expected_data, f"sample ix={ix} has data={data}, expected {expected_data}"
+
+    # If shuffling, there should be few matches
+    # The probability of k matches in a random permutation is ~1/(e*(k!))
+    if shuffle:
+        assert shuffle_matches < 10
+
```

Comment:
If shuffle is False, can we assert that the dataset is indeed not shuffled?

Reply:
This is accomplished within the loop, where every sample is assert'd to be the correct one (in-order)